feature(게시물 생성 페이지): 게시물 생성 페이지 기능 개발

Derek-94 opened the pull request:
이번 내용도 조금 자세하게 적습니다. 꼭 꼼꼼히 읽어주세요.
---
### 간략적인 구현 내용

- `ArticleCreatePage` 내부에서 일어나는 기능 구현.
- `title`, `hashtag`, `anonymous`, `contents` 내용을 콜백으로 넘김. 
- 게시물을 등록하려고 `submit` 을 누를때, `confirm` 메세지 노출.
- 게시물을 등록하거나 취소할때 `router` 를 사용한다.

---

### 구현 코드 설명.

1. 게시물을 새롭게 생성하는 것의 시작은 `ArticleCreatePageContainer` 입니다.

+) 이렇게 `Route` 를 통해서 해당 컨테이너에 접근합니다.

![image](https://user-images.githubusercontent.com/52649378/116973257-3752b980-acf7-11eb-8601-1db3a271a20d.png)

- 위 그림 28번째 줄에서 `/wirte` 경로를 설정해주었고, 이를 통해 접근합니다.

![image](https://user-images.githubusercontent.com/52649378/116974069-79302f80-acf8-11eb-8608-6e23cede3acb.png)

- 위 그림 12번째 줄에 보이는 `newArticle` 는, `presenter` 컴포넌트에서 얻어온 새로운 게시물 정보이다.

![image](https://user-images.githubusercontent.com/52649378/116774351-04aa8600-aa97-11eb-891f-e69923694668.png)

- 위와 같이 생겼다. 

- 다시, `container` 영역을 보면 9번째 줄에서 `api` 호출이 일어난다.

  - 즉, `presenter` 컴포넌트에서 얻어온 새로운 게시물 정보를 `POST` 로 등록한다.
 
- 그러고 나서 해당 게시물 `ArticleDetail` 페이지로 가는것은 **history의 push** 함수로 구현했다.

- 게시물이 등록되면 해당 게시물로 이동해야하기 때문에, 등록버튼이 눌리면 상세페이지로 넘어디=

---

### 이제 `presenter` 부분을 살펴보자.

![image](https://user-images.githubusercontent.com/52649378/116774443-90241700-aa97-11eb-9c42-3e38287c76a7.png)

- 9번째 줄을 제외하고 보자.

- 사용자가 입력하는 `title` , `hashtags` , `content` 내용을 `useState` 로 저장한다.

**사용자가 모두 입력을 했고, 이를 등록하는 과정은 아래와 같다.** 

![image](https://user-images.githubusercontent.com/52649378/116774490-d2e5ef00-aa97-11eb-8ad7-ae011230e537.png)

- 32번째 줄~ 40번째 줄은, `hashtags`를 쉼표 기준으로 자르는 과정이다.
- 여러개면 자르며, 사용자가 해쉬태그를 하나만 추가했으면 그것만 집어넣는다.
- `trim()` 함수를 통해 해쉬태그 앞뒤로 공백이 있으면 제거했다.
- **그렇게, 48번째 줄에서 콜백 함수로 새롭게 생성한 `article` 정보를 넘긴다.**

---

**사용자가 등록/취소 버튼을 눌렀을때 동작은 아래와 같다.**

![image](https://user-images.githubusercontent.com/52649378/116974196-ae3c8200-acf8-11eb-9198-9451442266a1.png)

- 사용자가 게시물을 등록하고자 하면, `onConfirmRegister` 함수를 통해, 먼저 **정말 등록할래?** 물어보고, `onSubmit()` 함수를 수행하여 등록한다.

 - 생각해보니까, 위에 등록하고 그 등록한 게시물 페이지로 들어가는게 여기 있어야 하나 싶다.

- 사용자가 취소버튼을 누르면, 59번째 줄 을 통해 만약 뭔가를 적었다면 **너 쓰던거 있는데 나갈거야?** 물어본다.

  - 여기서 나간다고 누르면, 이전 페이지로 넘어가야한다. 이거는 리액트 라우터로 구현했다. `history.goBack()` 함수를 사용했다.


- 모든 `input` 태그에는 `required` 속성을 넣어서, 사용자가 안적고 등록할라하면 **야 여기 안적었다**, 말해주게 설정했다.


---


# 꼭 코드단위로 리뷰 부탁드립니다. 🙏
# 확실하게 이해하셨으면 approve 해주세요. ✅
# 작은거라도 이상하면 여쭤봐주세요. 🤔

### 위 내용을 토대로해서 `presenter` 부분을 한번 참고하시면 이해가 빠를 것 같아요.


**저도 현주님도, 다연님도 FE 가 익숙치 않기에, 서로 협력하는 것이 필요합니다. 🎈** 

closes #19

Derek-94 committed:
feat(container): 게시물 생성 페이지 컨테이너 개발 시작.

- 구조 설계.
- eslint 적용.

Derek-94 committed:
feat(container): 게시물 생성 페이지 컨테이너 개발 시작.

- 구조 설계.
- eslint 적용.

Derek-94 committed:
refactor: prettier 적용 해제

Derek-94 committed:
refactor: prettier 적용 해제

Derek-94 committed:
feat(ArticleCreatePageContainer): 구조 변경.

- 컨테이너 컴포넌트 구조 변경.

Derek-94 committed:
feat(ArticleCreatePageContainer): 구조 변경.

- 컨테이너 컴포넌트 구조 변경.

Derek-94 committed:
fix: Fixed bug, that hashtags are not divided in ",".

- Setting hashtags are not working.
- Fixed bug.

Derek-94 committed:
fix: Fixed bug, that hashtags are not divided in ",".

- Setting hashtags are not working.
- Fixed bug.

Derek-94 committed:
feat(api): Update api calls when register articles.

- Fixed bug that not spliting hashtags in ",".
- Trimed blank of hashtags.

Derek-94 committed:
feat(api): Update api calls when register articles.

- Fixed bug that not spliting hashtags in ",".
- Trimed blank of hashtags.

Derek-94 committed:
feat(articleCreatePage): Add confirming message.

- When user register article, or cancel registering article, confirm
  pops up.
- Asks if user wrote sth in page and press cancel button.

Derek-94 committed:
feat(articleCreatePage): Add confirming message.

- When user register article, or cancel registering article, confirm
  pops up.
- Asks if user wrote sth in page and press cancel button.

Derek-94 committed:
feat(input): Add input tags required.

- When user tries to register article without typing title, hashtags,
  content, makes error.

Derek-94 committed:
feat(input): Add input tags required.

- When user tries to register article without typing title, hashtags,
  content, makes error.

dayonxxi committed:
feature(게시물 상세 조회): 게시물 상세 페이지 바디 컴포넌트 개발 (#15)



- ArticleBody부분 content, image, hashtags 띄우기

hyunju95 committed:
feature(게시판 게시글 목록 조회): React Router 도입 (#31)

* React Router 도입
* Header / Body  / AuthorInfo  / Comment Area → Article Detail Component로 통합

Derek-94 committed:
feat(component): Switch component.

- Change ArticleCreatePage component to ArticleCreatePageContainer
  component.

dayonxxi commented on src/Article/presentational/ArticleCreatePage.tsx:
```diff
@@ -20,41 +20,59 @@ const ArticleCreatePage = (/* {onRegisterArticle}: ArticleCreatePageProps */) =>
-    const $anonymous = document.querySelector(
-      '.article-create-anonymous'
-    ) as HTMLInputElement;
+    const $anonymous = document.querySelector('#anonymous') as HTMLInputElement;
```
anonymous라는 id에 일치하는 항목이 있다면 그 값을 return시켜주는..?코드인가요??
querySlector의 사용법을 읽어봤는데도 잘 이해가 안되네요ㅜㅠㅠ

Derek-94 replied:
엇 `querySelector` 라는 함수는, DOM을 조작하기 위한 함수에요.

마치 `getElementById` , `getElementByClass` 처럼, `HTML` 에 존재하는 태그들에 접근해서, 거기 값이나.. 필요한
값들을 가져오기위한 접근자? 로 보셔도 될것 같아요.

[여기](https://ko.javascript.info/searching-elements-dom)에 보시면 진짜 자세하게, **정말정말 쉽게** 알려주고있으니까

꼭! 읽어봐주세용!

dayonxxi replied:
링크들어가서 읽어보고 완벽하게 이해했습니다!

dayonxxi commented on src/Article/presentational/ArticleCreatePage.tsx:
```diff
@@ -20,41 +20,59 @@ const ArticleCreatePage = (/* {onRegisterArticle}: ArticleCreatePageProps */) =>
-    ) as HTMLInputElement;
+    const $anonymous = document.querySelector('#anonymous') as HTMLInputElement;
+    let modifiedHashTags: string[] = [];
+    if (newHashtags.includes(',')) {
+      modifiedHashTags = newHashtags.split(',');
+      modifiedHashTags = modifiedHashTags.map((splitHashtag) =>
+        splitHashtag.trim()
+      );
```
어제 설명들으면서 느꼈던건데 hashtag 에 앞뒤 공백이 있을 경우 지워주는 부분 너무 좋은 것 같아요!!!

Derek-94 replied:
제가 해쉬태그를 공백을 많이 넣어서 경험에 기반했네요! 감사합니다 😀

dayonxxi commented on src/Article/presentational/ArticleCreatePage.tsx:
```diff
@@ -20,41 +20,59 @@ const ArticleCreatePage = (/* {onRegisterArticle}: ArticleCreatePageProps */) =>
+  };
+
+  const onConfirmRegister = (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
```
아무것도 입력되지 않았을 때 글이 등록되는 것을 막아주는 코드로 이해했는데 맞을까요?

Derek-94 replied:
아 `e.preventDefault()` 함수는 음 새로고침을 방지해줍니다!

가령, `a` 태그는 클릭이 되면 새로운 창으로 이동시키고.., `form` 태그는 뭔가 제출하는? submit 이벤트가 일어나죠!
그럴때 보통 새로고침이 되면서 해당 작업이 수행되는데, `e.preventDefault()` 함수는 이 새로고침을 방지만 하고, 다른 기능은 정상적으로 수행되게 해주는 함수입니다!

[여기](https://programming119.tistory.com/100)에 자세하게 설명이 나와있네요 :)

dayonxxi replied:
요 부분도 링크 들어가서 읽어보고 완벽히 이해했습니다 ㅎㅎ

dayonxxi commented on src/Article/presentational/ArticleCreatePage.tsx:
```diff
@@ -20,41 +20,59 @@ const ArticleCreatePage = (/* {onRegisterArticle}: ArticleCreatePageProps */) =>
+
+  const onConfirmRegister = (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    if (window.confirm('게시물을 등록하시겠습니까?')) {
+      onSubmit();
+    }
```
'등록했을 때 등록한 글의 페이지로 이동해야하는지' 라고 적어두신거 읽었는데 
게시물 목록으로 이동하게 되면 자신이 작성한 글이 게시판에 새로 업로드 된 것도 확인 가능하니까 게시물목록 페이지로 이동시키는건 어떨까 생각해봤습니다!

Derek-94 replied:
음 이건 현주님까지 의견을 모아봐야겠네요!

velog 같은 경우에는 새로운 블로그 글을 등록하면, 그 등록된 페이지가 잘 되었는지, 내가 쓴 글이 등록된 화면으로 
넘어가면서 보여주거든요. 

이게 제가 그 방식에 익숙해서 그게 더 좋겠다 생각할 수도 있긴해요!

일단은 @hyunju95 님 의견도 궁금합니다!

hyunju95 replied:
엇 저도 이 부분은 석기님 의견이랑 동일한데, 글을 입력하고 내용이 정상적으로 등록 되었는지 바로 확인해보고 싶은 경우가 많아서 게시물 상세 조회 화면으로 가는 게 더 좋을 것 같습니다

dayonxxi replied:
그렇담 저도 그 방법이 좋을 것 같아요!

dayonxxi commented on src/Article/presentational/ArticleCreatePage.tsx:
```diff
@@ -67,6 +85,7 @@ const ArticleCreatePage = (/* {onRegisterArticle}: ArticleCreatePageProps */) =>
             type="text"
             placeholder="해시태그를 쉼표를 기준으로 입력하세요..."
             onChange={onChangeHashtags}
+            required
```
이 required가 있으면 해시태그를 입력하지 않았을 때 글 등록이 안되는건가요??
그런거라면 제목과 글에는 required가 필요할 것 같은데 해시태그는 없어도 되지 않을까 생각됩니다!

Derek-94 replied:
엇 그런가요? 해쉬태그.. 안적어도 그냥 되게 할까요?? 

생각해보니까 해쉬태그.. 굳이 안써도 되긴 할 것 같아요. 해쉬태그는 뭐 그냥 간단한거 적는거니까요.

전 찬성입니다! @hyunju95 님 의견 공유 부탁드릴게요!

hyunju95 replied:
해쉬태그는 없어도 괜찮을 것 같네요!

Derek-94 replied:
넵 알겠습니다! 해쉬태그가 없어도 게시물이 등록되게 바꿔보겠습니다 :)

dayonxxi reviewed (commented):
자세히 설명 적어주셔서 코드 이해하기 훨씬 편했습니다! 수고하셨어요 :)

Derek-94 committed:
feat(router): Update router function, in canceling register article.

- Used history.goBack function.

Derek-94 committed:
Merge branch 'feature/19' of https://github.com/dev-floor/untitled-client into feature/19

Derek-94 committed:
feat(router): Add router funtions

- Go back when cancel registering article
- Go foward to article Detail when register article

hyunju95 commented on src/Article/container/ArticleCreatePageContainer.tsx:
```diff
@@ -0,0 +1,23 @@
+
+  return <ArticleCreatePage onRegisterArticle={onRegisterArticle} history = {history} />;
+};
+
```
해당 부분에선 Route의 Props중 history만 사용하는 것으로 보이는데 MatchParams interface나 Props에 match가 선언이 된 이유가 있을까요?

Derek-94 replied:
음 이거는 혹시몰라서 남겨두었어요! 

게시물 생성하는 페이지가 지금은 `~/write` 이렇게만 `Route` 가 정해졌는데, 이를 어떻게 수정할지..

또한 생성하는 페이지가 보통 어떤 `URL`을 가질지 모르겠더라구요. 혹시 나중에 인자로 받은 `match` 가 쓰일 수도 있겠다, 싶어서 냄겨두었습니다. 🤔

hyunju95 reviewed (commented):
리뷰 남겼습니다. 수고하셨습니다!

Derek-94 committed:
fix: Make hashtags not required component.

- Remove required method in hashtag input tag.
- Can set hashtags in empty string, when user doesn't input hashtags.

Derek-94 reviewed (commented):
@dayonxxi 님께서 피드백 주신 내용 반영했습니다.

다만, 몇가지 상의할 내용이 있어서 일단 제가 잠깐 stop 요청드릴게요.

일단은 @hyunju95 님 코드리뷰만 부탁드리고, approve는 피해주세요! 
회의때 이상한 사항? 이상 현상 설명드릴게요.

Derek-94 commented:
> @dayonxxi 님께서 피드백 주신 내용 반영했습니다.
> 
> 다만, 몇가지 상의할 내용이 있어서 일단 제가 잠깐 stop 요청드릴게요.
> 
> 일단은 @hyunju95 님 코드리뷰만 부탁드리고, approve는 피해주세요!
> 회의때 이상한 사항? 이상 현상 설명드릴게요.

근데 현주님께서 이게 통합이 되고 CSS 작업이 들어가야한다 하시면, 먼저 그럼 `approve` 해주시고, 작업하셔도될것같습니다.

hyunju95 commented:
피드백 반영부분확인했습니다!

hyunju95 reviewed (approved):
피드백반영부분확인했습니다 수고하셨습니다!
